### PR TITLE
Add chain benchmarks

### PR DESCRIPTION
**Pull request by paulhauner**
## Issue Addressed

NA

## Proposed Changes

Implements:

- Per slot/block/epoch transitions from just prior to the `v0.1` spec.
- A testing suite for the `BeaconChain` (`test_harness`)
- Misc modifications to other crates to support above changes.
- A `eth2/attester` crate which provides the state-machine-style logic for an attester. It is very similar to the `eth/block_producer` crate.
- Modify `eth2/block_producer` where the `BeaconChain` API has changed. Ripple these changes into the `validator_client` binary.
- Add a `get_deserialized` method to the beacon_block and beacon_state stores. This makes it a little cleaner if you're looking for a fully-deserialized instance instead of a "reader" that might/might not be reading directly from SSZ.
- In the `eth2/types` and `beacon_node/beacon_chain` dirs, ensure that all impls live in the same file as their struct.
 - Removes `eth2/validator_shuffing` as it has been replaced by `BeaconState::get_shuffing`.

## Additional Info

I have left the `beacon_chain/src/lmd_ghost.rs` file for minimal disruption to @AgeManning. Age, I assume you'll move this into `beacon_chain.rs` when you're ready.

The implementation here is pre-`v0.1` spec release. Considering we'll need to refactor to catch up to `v0.1`, I've left the `eth2/types/beacon_chain.rs` file is a bit of a mess with a variety of `...Error`s and a lack of detailed comments. I'll clean all this up in the `v0.1` update PR, with the specific goal of reducing the number of different `...Error` types.

There are a few dangling functions here, specifically around validator management (adding/exiting validators, etc). They will be included in the `v0.1` update to reduce double implementing.

My recommendation for this PR is to review for code-structure and consistency and avoid looking too deeply into the spec implementation, as it will be soon updated.

### CODE REVIEW

File: beacon_node/beacon_chain/test_harness/src/validator/validator.rs [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+    DidNotProduce(AttestationPollOutcome),
+    PollError(AttestationPollError),
+}
+
+/// A `BlockProducer` and `Attester` which sign using a common keypair.
+///
+/// The test validator connects directly to a borrowed `BeaconChain` struct. It is useful for
+/// testing that the core proposer and attester logic is functioning. Also for supporting beacon
+/// chain tests.
+pub struct TestValidator {
```
Comment (AgeManning) [inline beacon_node/beacon_chain/test_harness/src/validator/validator.rs]: rename to `validator` or file to `test_validator`?

File: eth2/attester/src/test_utils/signer.rs [hunk 2]
```diff
@@ -0,0 +1,31 @@
+use crate::traits::Signer;
+use std::sync::RwLock;
+use types::{Keypair, Signature};
+
+/// A test-only struct used to simulate a Beacon Node.
+pub struct TestSigner {
```
Comment (AgeManning) [inline eth2/attester/src/test_utils/signer.rs]: `signer` or `test_signer.rs`

File: eth2/block_producer/src/test_utils/epoch_map.rs [hunk 3]
```diff
@@ -0,0 +1,27 @@
+use crate::{DutiesReader, DutiesReaderError};
+use std::collections::HashMap;
+
+pub struct TestEpochMap {
```
Comment (AgeManning) [inline eth2/block_producer/src/test_utils/epoch_map.rs]: `EpochMap` or `test_epoch_map`

File: eth2/block_producer/src/test_utils/signer.rs [hunk 4]
```diff
@@ -0,0 +1,35 @@
+use crate::traits::Signer;
+use std::sync::RwLock;
+use types::{Keypair, Signature};
+
+/// A test-only struct used to simulate a Beacon Node.
+pub struct TestSigner {
```
Comment (AgeManning) [inline eth2/block_producer/src/test_utils/signer.rs]: signer or test_signer.rs

File: validator_client/src/block_producer_service/grpc.rs [hunk 5]
```diff
@@ -1,21 +1,45 @@
...
 use protos::services_grpc::BeaconBlockServiceClient;
 use ssz::{ssz_encode, Decodable};
-use types::{BeaconBlock, BeaconBlockBody, Eth1Data, Hash256, Signature};
+use std::sync::Arc;
+use types::{BeaconBlock, BeaconBlockBody, Eth1Data, Hash256, PublicKey, Signature};
 
-impl BeaconNode for BeaconBlockServiceClient {
+/// A newtype designed to wrap the gRPC-generated service so the `BeaconNode` trait may be
+/// implemented upon it.
+pub struct BeaconBlockGrpcClient {
```
Comment (AgeManning) [inline validator_client/src/block_producer_service/grpc.rs]: `grpc` or `beacon_block_grpc_client.rs`

Comment (AgeManning) [reply validator_client/src/block_producer_service/grpc.rs]: More about project consistency than anything else. See gitter.

File: validator_client/src/block_producer_service/service.rs [hunk 6]
```diff
@@ -1,15 +1,17 @@
...
+use block_producer::{
+    BeaconNode, BlockProducer, DutiesReader, PollOutcome as BlockProducerPollOutcome, Signer,
+};
 use slog::{error, info, warn, Logger};
+use slot_clock::SlotClock;
 use std::time::Duration;
 
-pub struct BlockProducerService<T: SlotClock, U: BeaconNode> {
-    pub block_producer: BlockProducer<T, U>,
+pub struct BlockProducerService<T: SlotClock, U: BeaconNode, V: DutiesReader, W: Signer> {
```
Comment (AgeManning) [inline validator_client/src/block_producer_service/service.rs]: Similar to other comments

File: validator_client/src/duties/duties_map.rs [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+use std::collections::HashMap;
+use std::sync::RwLock;
+
+/// The information required for a validator to propose and attest during some epoch.
+///
+/// Generally obtained from a Beacon Node, this information contains the validators canonical index
+/// (thier sequence in the global validator induction process) and the "shuffling" for that index
+/// for some epoch.
+#[derive(Debug, PartialEq, Clone, Copy, Default)]
+pub struct EpochDuties {
```
Comment (AgeManning) [inline validator_client/src/duties/duties_map.rs]: rename the file to epoch_duties.rs

File: beacon_node/beacon_chain/test_harness/src/validator/validator.rs (hunk 1 above)
Comment (paulhauner) [reply beacon_node/beacon_chain/test_harness/src/validator/validator.rs]: `Validator` clashes with `types::Validator` and `test_validator.rs` looks like a test. I'll try figure something out, lemme know if you have any ideas.

File: beacon_node/beacon_chain/test_harness/src/harness.rs [hunk 8]
```diff
@@ -0,0 +1,233 @@
...
+use std::sync::Arc;
+use types::{BeaconBlock, ChainSpec, FreeAttestation, Keypair, Validator};
+
+/// The beacon chain harness simulates a single beacon node with `validator_count` validators connected
+/// to it. Each validator is provided a borrow to the beacon chain, where it may read
+/// information and submit blocks/attesations for processing.
+///
+/// This test harness is useful for testing validator and internal state transition logic. It
+/// is not useful for testing that multiple beacon nodes can reach consensus.
+pub struct BeaconChainHarness {
```
[the remark this answers is not in the document]
Comment (paulhauner) [reply beacon_node/beacon_chain/test_harness/src/harness.rs]: I've been trying to ensure that we don't have naming collisions when we start dealing with `ShardChain`s.

[the remark this answers is not in the document]
Comment (paulhauner) [reply beacon_node/beacon_chain/test_harness/src/harness.rs]: But, that's probably best done at a higher level. I'll rename to `Harness`.

File: beacon_node/beacon_chain/test_harness/src/validator/validator.rs (hunk 1 above)
Comment (AgeManning) [reply beacon_node/beacon_chain/test_harness/src/validator/validator.rs]: `validator_test` ?

File: eth2/attester/src/test_utils/signer.rs (hunk 2 above)
Comment (AgeManning) [reply eth2/attester/src/test_utils/signer.rs]: As with validator, maybe `SignerTest`

**Comment by paulhauner** [comment]
Summary:

I changed some names away from `Test...` because its easy to confuse it with test files. Also, there's probably a more descriptive word than "Test" that suits the job better.

Changelog:

`beacon_chain`:

- `TestValidator` -> `ValidatorHarness`
- `BeaconChainHarness` untouched (because there's now another harness).
- `TestSigner` -> `LocalSigner` cause it stores keys locally.

`attester` and `block_producer`:
- `TestSigner` -> `LocalSigner` cause it stores keys locally.
- `TestBeaconNode` -> `SimulatedBeaconNode`.

`validator_client`:
- All suggestions implemented as stated.

**Pull request merged**